File: sources/Services/Storage/RamCacheDB/ramcachedb.cpp
#include "ramcachedb.h"

RamCacheDB::RamCacheDB(DependencyInjectionManager* dim)
{ 
    this->confs = dim->get<Confs>();
    this->log = dim->get<ILogger>()->getNamedLoggerP("RamCacheDBStorage");

    confs->listenA("DbDirectory", [&](DynamicVar value)
    {
        this->log->info("Database directory: " + value.getString());
        this->dataDir = value.getString();
        this->load();
    });

    confs->listenA("RamCacheDbDumpIntervalMs", [&](DynamicVar value)
    {
        this->log->info("Dump interval setted to: " + value.getString() + "ms");
        this->dumpIntervalMs = value.getInt();
    });


    dumpThread = new thread([=](){
        int waitingtotalTimeMs=0;
        while (continueRunning)
        {
            if (waitingtotalTimeMs > dumpIntervalMs) //5 minutes
            {
                dump();
                waitingtotalTimeMs=0;
            }
            this_thread::sleep_for(chrono::milliseconds(50));
            waitingtotalTimeMs+=50;
        }
    });
}

RamCacheDB::~RamCacheDB()
{
    continueRunning = false;
    this->dumpThread->join();
    dump();
}

RamCacheDBItem* RamCacheDB::_scrollTree(string name, RamCacheDBItem &curr, bool readOnly)
{
    auto currName = name;
    string remainingName = "";
    if (name.find('.') != string::npos)
    {
        currName = name.substr(0, name.find('.'));
        remainingName = name.substr(name.find('.')+1);
    }

    if (curr.childs.count(currName) == 0)
    {
        if (readOnly)
            return nullptr;

        curr.childs[currName] = RamCacheDBItem();
        curr.childs[currName].imediateName = currName;
        curr.childs[currName].fullName = curr.fullName == "" ? currName : curr.fullName + "." + currName;
        curr.childs[currName].parent = &curr;

    }

    if (remainingName == "")
    {
        return &curr.childs[currName];
    }

    return _scrollTree(remainingName, curr.childs[currName], readOnly);
    
}

void RamCacheDB::set(string name, DynamicVar v)
{
    dblocker.lock();
    auto item = _scrollTree(name, root, false);
    if (item != nullptr)
    {
        item->value = v;
        pendingChanges = true;
    }
    else
    {
        dblocker.unlock();
        throw runtime_error("Error trying to set value on RamCacheDB, item not found: " + name);
    }

    dblocker.unlock();
}



DynamicVar RamCacheDB::get(string name, DynamicVar defaultValue)
{
    auto ret = defaultValue;
    dblocker.lock();
    auto item = _scrollTree(name, root, true);
    if (item != nullptr)
        ret = item->value;
        
    dblocker.unlock();
    return ret;
}


//return only imediate childs, do not return subschilds (childs of childs). Return only imediate key name (the full key name should not be returned)
vector<string> RamCacheDB::getChilds(string parentName)
{
    vector<string> ret;
    
    dblocker.lock();
    auto item = _scrollTree(parentName, root, true);
    if (item != nullptr)
    {
        for (auto &c: item->childs)
            ret.push_back(c.first);
    }
    dblocker.unlock();

    return ret;
}

bool RamCacheDB::hasValue(string name)
{
    dblocker.lock();
    auto item = _scrollTree(name, root, true);
    dblocker.unlock();
    return item != nullptr;
}

void RamCacheDB::deleteValue(string name, bool deleteChildsInACascade)
{
    dblocker.lock();
    auto item = _scrollTree(name, root, false);
    if (item != nullptr)
    {
        item->value = DynamicVar();

        if (deleteChildsInACascade)
            item->childs.clear();

        if (item->childs.size() == 0){
            item->parent->childs.erase(item->imediateName);
        }
        pendingChanges = true;
    }
    dblocker.unlock();
}

void RamCacheDB::forEachChilds(string parentName, function<void(string, DynamicVar)> f)
{
    dblocker.lock();
    auto item = _scrollTree(parentName, root, true);
    if (item != nullptr)
    {
        for (auto &c: item->childs)
            f(parentName + "." + c.first, c.second.value);
    }
    dblocker.unlock();

}

future<void> RamCacheDB::forEachChilds_parallel(string parentName, function<void(string, DynamicVar)> f, ThreadPool *taskerForParallel)
{
    dblocker.lock();
    auto item = _scrollTree(parentName, root, true);
    if (item != nullptr)
    {
        for (auto &c: item->childs)
            taskerForParallel->enqueue([=](){f(parentName + "." + c.first, c.second.value); });
    }
    dblocker.unlock();

    std::promise<void> p;
    p.set_value();
    return p.get_future();
}

string RamCacheDB::dumpToString(RamCacheDBItem &current, string currentParentName)
{
    string ret = "";
    if (currentParentName != "")
        currentParentName = currentParentName + ".";
    for (auto &c: current.childs)
    {
        if (c.second.value.getString() !=  "")
            ret += currentParentName + c.first + "=" + c.second.value.getString() + "\n";

        ret += dumpToString(c.second, currentParentName +  c.first);
    }
    return ret;
}

void RamCacheDB::dump()
{
    if (!pendingChanges)
        return;

    pendingChanges = false;


    
    this->log->debug("Dumping database to disk");
    dblocker.lock();
    string fileText=dumpToString(root);
    dblocker.unlock();
    
    //create directory
    Utils::ssystem("mkdir -p \"" + dataDir+"\"");

    Utils::writeTextFileContent(dataDir + "/"+ DUMP_FILE_NAME, fileText);
    this->log->debug("Dumping database to disk finished");
}

string getDumpFileName();


void RamCacheDB::load()
{
    this->log->info("Loading database from disk");
    string fileText = Utils::readTextFileContent(getDumpFilePath());
    auto lines = Utils::splitString(fileText, "\n");
    dblocker.lock();
    for (auto &c: lines)
    {
        auto parts = Utils::splitString(c, "=");
        if (parts.size() == 2)
        {
            auto item = _scrollTree(parts[0], root, false);
            if (item != nullptr)
                item->value = parts[1];
            else
            {
                dblocker.unlock();
                throw runtime_error("RamCacheDB Error trying to loading a value from the disk, item not found: " + parts[0]);
            }
        }
    }
    dblocker.unlock();
    this->log->info("Loading database from disk finished");
}

string RamCacheDB::getDumpFilePath()
{
    return dataDir + "/"+ DUMP_FILE_NAME;
}

File: sources/services/storage/ramcachedb.go
package storage

import (
	"os"
	"path/filepath"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"rtonello/vss/sources/misc"
	"rtonello/vss/sources/misc/logger"
)

// ramNode represents a node in the in-memory tree storage.
type ramNode struct {
	imediateName string
	parent       *ramNode
	childs       map[string]*ramNode
	value        misc.DynamicVar
}

// RamCacheDB is a simple in-memory tree-based storage that mirrors the C++ RamCacheDB
// behavior: keys are dot-separated (e.g. "a.b.c"), GetChilds returns immediate child names
// (not full names) and ForEachChilds passes the full child key to the callback.
type RamCacheDB struct {
	mu             sync.RWMutex
	root           ramNode
	pendingChanges atomic.Bool

	// dump configuration
	dataDir        string
	dumpIntervalMs int
	stopCh         chan struct{}
	log            logger.INamedLogger
}

// NewRamCacheDBWithOptions creates and returns an initialized RamCacheDB with options.
// If dataDir is empty a default is used; if dumpIntervalMs <= 0 a default 5min is used.
func NewRamCacheDB(logger logger.ILogger, dataDir string, dumpIntervalMs int) IStorage {
	r := &RamCacheDB{}
	r.root.childs = make(map[string]*ramNode)
	r.root.imediateName = ""
	r.log = logger.GetNamedLogger("Storage::RamCacheDB")

	if dataDir == "" {
		r.dataDir = "./data/database"
	} else {
		r.dataDir = dataDir
	}
	if dumpIntervalMs <= 0 {
		r.dumpIntervalMs = 5 * 60 * 1000 // 5 minutes
	} else {
		r.dumpIntervalMs = dumpIntervalMs
	}

	r.stopCh = make(chan struct{})
	// load existing data
	if err := r.load(); err != nil {
		r.log.Error("Error loading RamCacheDB from disk: " + err.Error())
		r.log.Warning("Continuing with empty RamCacheDB.")
	}

	//load data from disk
	if err := r.load(); err != nil {
		r.log.Error("Error loading RamCacheDB from disk: " + err.Error())
		r.log.Warning("Continuing with empty RamCacheDB.")
	}

	go r.backgroundDumper()
	return r
}

// redirects to NewRamCacheDB with auto defined parameters
func NewRamCacheDB2(logger logger.ILogger) IStorage {
	return NewRamCacheDB(logger, "", 0)
}

// Close stops the background dumper and performs a final dump.
func (r *RamCacheDB) Finalize() {
	// close stopCh once (safe non-blocking check)
	select {
	case <-r.stopCh:
		// already closed
		return
	default:
		close(r.stopCh)
	}
}

func (r *RamCacheDB) backgroundDumper() {
	ticker := time.NewTicker(time.Duration(r.dumpIntervalMs) * time.Millisecond)
	defer ticker.Stop()
	for {

		select {
		case <-r.stopCh:
			r.mu.Lock()
			if r.pendingChanges.Load() {
				r.dump()
				r.pendingChanges.Store(false)
			}
			r.mu.Unlock()

			return
		case <-ticker.C:
			r.mu.Lock()
			if r.pendingChanges.Load() {
				r.dump()
				r.pendingChanges.Store(false)
			}
			r.mu.Unlock()
		}
	}
}

// helper: traverse tree, creating nodes if create=true
func (r *RamCacheDB) scrollTree(name string, curr *ramNode, readOnly bool) *ramNode {
	currName := name
	remainingName := ""
	if strings.Contains(name, ".") {
		indx := strings.Index(name, ".")
		currName = name[:indx]
		remainingName = name[indx+1:]
	}

	if curr.childs[currName] == nil {
		{
			if readOnly {
				return nil
			}

			curr.childs[currName] = &ramNode{
				imediateName: currName,
				parent:       curr,
				childs:       make(map[string]*ramNode),
				value:        misc.NewEmptyDynamicVar(),
			}

			//if curr.fullName != "" {
			//	curr.childs[currName].fullName = curr.fullName + "." + currName
			//} else {
			//	curr.childs[currName].fullName = currName
			//}

		}
	}

	if remainingName == "" {
		return curr.childs[currName]
	}

	return r.scrollTree(remainingName, curr.childs[currName], readOnly)
}

// Set stores value v at key name.
func (r *RamCacheDB) Set(name string, v misc.DynamicVar) {
	r.mu.Lock()
	defer r.mu.Unlock()
	item := r.scrollTree(name, &r.root, false)
	if item != nil {
		item.value = v
		r.pendingChanges.Store(true)
	}
}

// Get returns the stored value or defaultValue if not present.
func (r *RamCacheDB) Get(name string, defaultValue misc.DynamicVar) misc.DynamicVar {
	r.mu.RLock()
	defer r.mu.RUnlock()
	item := r.scrollTree(name, &r.root, true)
	if item != nil {
		return item.value
	}
	return defaultValue
}

// GetChilds returns immediate child names (not full names) for parentName.
func (r *RamCacheDB) GetChilds(parentName string) []string {
	ret := []string{}
	if parentName == "" {
		//get root childs
		r.mu.RLock()
		defer r.mu.RUnlock()
		for k := range r.root.childs {
			ret = append(ret, k)
		}
		return ret
	}

	r.mu.RLock()
	defer r.mu.RUnlock()
	item := r.scrollTree(parentName, &r.root, true)
	if item == nil {
		return []string{}
	}

	ret = make([]string, 0, len(item.childs))
	if item == nil || item.childs == nil {
		return []string{}
	}
	for k := range item.childs {
		ret = append(ret, k)
	}
	return ret
}

// HasValue returns true if the node for name exists (matches C++ semantics).
func (r *RamCacheDB) HasValue(name string) bool {
	r.mu.RLock()
	defer r.mu.RUnlock()
	item := r.scrollTree(name, &r.root, true)
	return item != nil
}

// DeleteValue removes the value at name; if deleteChildsInACascade is true, remove children as well.
func (r *RamCacheDB) DeleteValue(name string, deleteChildsInACascade bool) {
	r.mu.Lock()
	defer r.mu.Unlock()
	item := r.scrollTree(name, &r.root, true)
	if item == nil {
		return
	}
	// clear value
	item.value = misc.NewEmptyDynamicVar()
	if deleteChildsInACascade {
		item.childs = make(map[string]*ramNode)
	}
	if len(item.childs) == 0 && item.parent != nil {
		delete(item.parent.childs, item.imediateName)
	}
	r.pendingChanges.Store(true)
}

// ForEachChilds iterates immediate childs and calls f(fullChildName, value).
func (r *RamCacheDB) ForEachChilds(parentName string, f func(string, misc.DynamicVar)) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	item := r.scrollTree(parentName, &r.root, true)
	if item == nil || item.childs == nil {
		return
	}
	for k, child := range item.childs {
		full := k
		if parentName != "" {
			full = parentName + "." + k
		}
		f(full, child.value)
	}
}

// ForEachChildsParallel iterates immediate childs and calls f concurrently using a worker pool.
// Returns a channel that's closed when all work is done; caller can wait on it using <-done.
func (r *RamCacheDB) ForEachChildsParallel(parentName string, f func(string, misc.DynamicVar), concurrency int) <-chan struct{} {
	done := make(chan struct{})
	if concurrency < 1 {
		concurrency = 1
	}

	r.mu.RLock()
	item := r.scrollTree(parentName, &r.root, true)
	if item == nil || item.childs == nil {
		r.mu.RUnlock()
		close(done)
		return done
	}

	// gather tasks
	tasks := make([]struct {
		full string
		val  misc.DynamicVar
	}, 0, len(item.childs))
	for k, child := range item.childs {
		full := k
		if parentName != "" {
			full = parentName + "." + k
		}
		tasks = append(tasks, struct {
			full string
			val  misc.DynamicVar
		}{full: full, val: child.value})
	}
	r.mu.RUnlock()

	var wg sync.WaitGroup
	taskCh := make(chan struct {
		full string
		val  misc.DynamicVar
	})

	// start workers
	for i := 0; i < concurrency; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for t := range taskCh {
				f(t.full, t.val)
			}
		}()
	}

	// feed tasks
	go func() {
		for _, t := range tasks {
			taskCh <- t
		}
		close(taskCh)
		wg.Wait()
		close(done)
	}()

	return done
}

// dump expects r.mu to be locked. It writes the dump file to disk and returns any error.
func (r *RamCacheDB) dump() {
	r.log.Trace("Dumping RamCacheDB to disk...")
	text := r.dumpToStringLocked(&r.root, "")
	// ensure directory
	if err := os.MkdirAll(r.dataDir, 0o755); err != nil {
		r.log.Error("Error dumping RamCacheDB to disk: " + err.Error())
		return
	}
	path := filepath.Join(r.dataDir, "RamCacheDb.dump.txt")
	ret := os.WriteFile(path, []byte(text), 0o644)
	if ret != nil {
		r.log.Error("Error dumping RamCacheDB to disk: " + ret.Error())
		return
	}
	r.log.Trace("Finished dumping RamCacheDB to disk.")
}

func (r *RamCacheDB) dumpToStringLocked(current *ramNode, currentParentName string) string {
	ret := ""
	prefix := ""
	if currentParentName != "" {
		prefix = currentParentName + "."
	}
	for k, c := range current.childs {
		if c.value.GetString() != "" {
			ret += prefix + k + "=" + c.value.GetString() + "\n"
		}
		ret += r.dumpToStringLocked(c, prefix+k)
	}
	return ret
}

// Load reads the dump file and populates the in-memory cache. Existing nodes are preserved/extended.
func (r *RamCacheDB) load() error {
	path := filepath.Join(r.dataDir, "RamCacheDb.dump.txt")
	b, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return err
	}
	text := string(b)
	lines := strings.Split(text, "\n")
	r.mu.Lock()
	defer r.mu.Unlock()
	for _, line := range lines {
		if strings.TrimSpace(line) == "" {
			continue
		}
		idx := strings.Index(line, "=")
		if idx < 0 {
			continue
		}
		key := line[:idx]
		val := line[idx+1:]
		item := r.scrollTree(key, &r.root, false)
		if item != nil {
			item.value = misc.NewDynamicVar(val)
		}
	}
	r.pendingChanges.Store(false)
	return nil
}
